### PR TITLE
Use yaml-rust 0.4

### PR DESCRIPTION
**Pull request by vroland**
As yaml-rust 0.4 got publisched on crates.io recently, it would be nice to use it for serde-yaml. There are some some fixes (especially for the emitter) I'd really like use with serde.

### CODE REVIEW

**Comment by dtolnay** [comment]
@vroland I haven't looked through the yaml-rust changes in detail so I need your advice -- is this safe to publish as 0.7.3 of serde-yaml or does the input/output change in a way that requires a major version from us?

File: tests/test_serde.rs [hunk 1]
```diff
@@ -94,7 +94,7 @@ fn test_map() {
     let yaml = unindent(r#"
         ---
         x: 1
-        y: 2"#);
+        "y": 2"#);
```
Comment (ignatenkobrain) [inline tests/test_serde.rs]: I think this is wrong and there is some bug... @dtolnay

Comment (dtolnay) [reply tests/test_serde.rs]: This is from https://github.com/chyh1990/yaml-rust/pull/72.

Comment (vroland) [reply tests/test_serde.rs]: Yes, this is probably an issue with the emitter. But this is more of a cosmetic issue, as it is still valid YAML with the same meaning.

**Comment by vroland** [comment]
As far as I am aware these breaking changes are only present in parser event handling, which is not exposed by serde-yaml. But ScanError struct contains an additional field "mark". If this is a problem we need a higher version.

File: tests/test_serde.rs (hunk 1 above)
Comment (hoodie) [reply tests/test_serde.rs]: this is intended to keep any other yaml1.1 (not 1.2) parser from interpreting this as a [boolean](http://yaml.org/type/bool.html). I am currently unaware whether this ought to be treated differently when the value is a key.

**Comment by Rufflewind** [comment]
It looks v0.4 changes the rendering for certain cases.  Previously, I got:

```
blah: 
  {a: 1, b: 2, c: 3}: 1.2345
```

Now I get:

```
blah:
  ? "a": 1
    b: 2
    c: 3
  : 1.2345
```

I found it confusing that the major version did not increase.  (Also, the quoting around just `"a"` but none of the other keys looks weird.)

**Comment by hoodie** [comment]
@Rufflewind: semver allows this. Below 1.0 every 0.X version may contain breaking changes.
The emitter has been replaced in this version. I was not aware of this behavior before, but it appears to be valid yaml. The 0.3.x version used to emit invalid yaml in certain edgecases. Seems we should look into this again after all, but at least it's parsable now :P

File: tests/test_serde.rs (hunk 1 above)
Comment (chyh1990) [reply tests/test_serde.rs]: I have tried the yaml parser in ruby 2.3, 'Y', 'y', 'N', 'n' is not interpreted as boolean in both keys and values. But 'YES', 'yes', 'NO', 'no' is boolean in ruby. PyYAML also works like this.

I know it is violating YAML spec,  the using 'y' in key is a common case.

Comment (hoodie) [reply tests/test_serde.rs]: the ruby parses an pyyaml are all 1.1 parsers

**Pull request merged**